subxt/src/storage/storage_key.rs

```diff
@@ -0,0 +1,232 @@
+use super::{
+    storage_address::StaticStorageKey,
+    utils::{strip_concat_hash_bytes, strip_storage_addess_root_bytes},
+};
+use crate::{
+    dynamic::DecodedValueThunk,
+    error::{Error, StorageAddressError},
+    metadata::{DecodeWithMetadata, Metadata},
+    utils::{Encoded, Static},
+};
+use scale_encode::EncodeAsType;
+use subxt_metadata::StorageHasher;
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType>;
+
+    /// Attempts to decode the StorageKey from a whole storage address.
+    /// The key/keys can only be recovered if all hashers are concat-style hashers.
+    /// Example: Imagine The `StorageKey` is a tuple (A,B) and the hashers are: [Blake2_128Concat, Twox64Concat].
+    /// Then the memory layout of the storage key is:
+    /// ```txt
+    /// | 8 bytes pallet hash | 8 bytes entry hash | 16 bytes hash of A | ... bytes of A | 8 bytes hash of B | ... bytes of B |
+    /// ```
+    /// Returns None, if any of the hashers is not a concat-style hasher.
```

Comment:
What if one of the keys is a concat hasher and one isn't?

I'd assume that the codegen for that _could_ allow eg `((), AccountId32)` to be decoded from storage keys like `blake2_256, blake2_256_concat`? I wonder how much complexity supporting something like this would add, if not already possible?

Reply:
Perhaps more to the point: since this is a trait, it offers the possibility to return `None` if decoding can't succeed, but the implementations get to decide when to use this, so the comment should probably be more like "this is expected to return None when X".

And this gets me thinking; is `None` just an error; ie if you try to decode storage keys into something of the wrong shape, the decoding will fail basically (which in most places in Subxt leadds to a `DecodeError` or something). So is there a point in returning `Option<Result<>>` and not just returning `Result` and having a suitable error for when decoding fails if none already exists? Option makes sense for streams etc that can return `None` under a normal circumstance, but I can't think what it would mean here.